### PR TITLE
feat(cache/unstable): `TtlCache`

### PR DESCRIPTION
**Pull request by lionel-rowe**
Closes https://github.com/denoland/std/issues/5669

### CODE REVIEW

**Comment by iuioiua** [comment]
Please see https://github.com/denoland/std/issues/4608#issuecomment-2277606460

File: cache/ttl_cache.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+import type { MemoizationCache } from "./memoize.ts";
+import { delay } from "@std/async/delay";
+
+/**
+ * Time-to-live cache.
+ *
+ * Automatically removes entries once the configured amount of time elapses. If
+ * the values themselves are promises, the countdown starts from when they
+ * resolve.
```
Comment (iuioiua) [inline cache/ttl_cache.ts:10]: Hm... I'm not sure that promises should get any special treatment... @oles, do you have any preferences?

File: cache/ttl_cache.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ * @typeParam K The type of the cache keys.
+ * @typeParam V The type of the cache values.
+ *
+ * @example Usage
+ * ```ts
+ * import { TtlCache } from "@std/cache";
```
Comment (iuioiua) [inline cache/ttl_cache.ts:17]:
```suggestion
 * import { TtlCache } from "@std/cache/ttl-cache";
```

File: cache/ttl_cache.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * import { assertEquals } from "@std/assert";
```
Comment (iuioiua) [inline cache/ttl_cache.ts:18]:
```suggestion
 * import { assertEquals } from "@std/assert/equals";
```

File: cache/ttl_cache.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+   *
+   * @example Usage
+   * ```ts
+   * import { TtlCache } from "@std/cache";
+   * import { assertEquals } from "@std/assert";
+   * const cache = new TtlCache(1000);
+   * assertEquals(cache.defaultTtl, 1000);
+   * ```
+   */
+  defaultTtl: number;
```
Comment (iuioiua) [inline cache/ttl_cache.ts:42]: I think it's best that we don't expose any properties unless we have a real good reason to. I think it's fine if this is private.
```suggestion
  #defaultTtl: number;
```

File: cache/ttl_cache.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+   * import { TtlCache } from "@std/cache";
+   * import { assertEquals } from "@std/assert";
+   *
+   * const cache = new TtlCache<string, number>(1000);
+   *
+   * cache.set("a", 1);
+   * assertEquals(cache.get("a"), 1);
+   * ```
+   */
+  override set(key: K, value: V, customTtl?: number): this {
```
Comment (iuioiua) [inline cache/ttl_cache.ts:74]:
```suggestion
  override set(key: K, value: V, customTtl = this.#defaultTimeout): this {
```

File: cache/ttl_cache.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    this.#deleters.get(key)?.ac.abort();
+    const ac = new AbortController();
+
+    this.#deleters.set(key, {
+      promise: (async () => {
+        if (value instanceof Promise) {
+          await value;
+        }
+
+        await delay(customTtl ?? this.defaultTtl);
```
Comment (iuioiua) [inline cache/ttl_cache.ts:86]: I wonder whether we really need this `delay()` dependency. Might we be just fine using [`AbortSignal.timeout()`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout_static)?

File: cache/ttl_cache_test.ts
Comment (iuioiua) [inline cache/ttl_cache_test.ts:1]: Can you please sprinkle in some use of `.get()`, so we know that the returned values are correct?

**Review by iuioiua — CHANGES_REQUESTED** [review]

File: cache/ttl_cache.ts (hunk 1 above)
Comment (kt3k) [reply cache/ttl_cache.ts:10]: That sounds a good point. I don't see the rationale for the special handling of `Promise` very well.

Also this seems causing uncaught error when the promise is rejected. That is probably inconvenient for most users.

File: cache/ttl_cache.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ * assertEquals(cache.size, 0);
+ * ```
+ */
+export class TtlCache<K, V> extends Map<K, V>
+  implements MemoizationCache<K, V> {
+  #defaultTtl: number;
+  #timeouts = new Map<K, number>();
+
+  /**
+   * Construct a new `TtlCache`.
```
Comment (iuioiua) [inline cache/ttl_cache.ts:32]:
```suggestion
   * Constructs a new instance.
```

**Review by iuioiua — CHANGES_REQUESTED** [review]
PTAL, @kt3k.

File: cache/ttl_cache_test.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  await t.step("default TTL (passed in constructor)", () => {
+    using time = new FakeTime(0);
+
+    const cache = new TtlCache<number, string>(10);
+
+    cache.set(1, "one");
+    cache.set(2, "two");
+
+    time.now = 1;
+    assertEntries(cache, [[1, "one"], [2, "two"]]);
```
[the remark this answers is not in the document]
Comment (lionel-rowe) [reply cache/ttl_cache_test.ts:38]: I added `assertEntries` and `UNSET` to make sure `has`, `get`, etc. are hit as well as just the iterable entries themselves (`UNSET` checks `has` is false)

**Review by iuioiua — APPROVED** [review]
LGTM! Thank you.

File: cache/ttl_cache.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+   *
+   * @example Usage
+   * ```ts no-assert
+   * import { TtlCache } from "@std/cache/ttl-cache";
+   * console.log(new TtlCache<string, number>(1000)); // TtlCache
+   * ```
+   */
+  override get [Symbol.toStringTag](): string {
+    return this.constructor.name;
+  }
```
Comment (iuioiua) [inline cache/ttl_cache.ts:43]: Please add a test

**Review by kt3k — APPROVED** [review]
LGTM when Asher's point addressed

**Review by iuioiua — APPROVED** [review]
On second thought, I removed `[Symbol.toStringTag]()`. We don't implement this anywhere else in `std`. Either way, thank you for this PR!

**Review by jsejcksn — COMMENTED** [review]
It might be useful to document the valid range for TTL values in this implementation. Because it defers to `setTimeout`, this will be implementation-dependent (see link to spec in previous comment), but testing can be performed to inform a suggestion for Deno at least. Here is an example demonstrating that the current build of V8 in Deno only properly handles unsigned 32-bit integers:

`settimeout_range.test.ts`:
```lang-ts
import { AssertionError } from "jsr:@std/assert@1.0.3/assertion-error";

console.log("Deno.version", Deno.version);

async function assertIntervalIsScheduledAfter1Millisecond(
  interval: number,
): Promise<void> {
  let timestamp: number | undefined;
  const expectedTimeStamp = Date.now() + interval;

  const timerId = setTimeout(() => {
    timestamp = Date.now();
  }, interval);

  await new Promise((resolve) => setTimeout(resolve, 1));

  if (timestamp) {
    throw new AssertionError(
      `The callback completed ${
        expectedTimeStamp - Date.now()
      } ms earlier than expected.`,
    );
  }

  clearTimeout(timerId);
}

Deno.test("setTimeout(): Schedules intervals…", async (t) => {
  await t.step(
    "less than 2 ** 31",
    () => assertIntervalIsScheduledAfter1Millisecond(2 ** 31 - 1),
  );

  await t.step(
    "greater than or equal to 2 ** 31",
    () => assertIntervalIsScheduledAfter1Millisecond(2 ** 31),
  );
});
```

Test output:

```txt
% deno test settimeout_range.test.ts
------- pre-test output -------
Deno.version { deno: "1.46.1", v8: "12.9.202.2-rusty", typescript: "5.5.2" }
----- pre-test output end -----
running 1 test from ./settimeout_range.test.ts
setTimeout(): Schedules intervals… ...
  less than 2 ** 31 ... ok (3ms)
  greater than or equal to 2 ** 31 ... FAILED (1ms)
setTimeout(): Schedules intervals… ... FAILED (due to 1 failed step) (4ms)

 ERRORS 

setTimeout(): Schedules intervals… ... greater than or equal to 2 ** 31 => ./settimeout_range.test.ts:34:10
error: AssertionError: The callback completed 2147483647 ms earlier than expected.
		throw new AssertionError(
		      ^
    at assertIntervalIsScheduledAfter1Millisecond (file:///Users/deno/settimeout_range.test.ts:18:9)
    at eventLoopTick (ext:core/01_core.js:213:9)
    at async innerWrapped (ext:cli/40_test.js:191:5)
    at async exitSanitizer (ext:cli/40_test.js:107:27)
    at async Object.outerWrapped [as fn] (ext:cli/40_test.js:134:14)
    at async TestContext.step (ext:cli/40_test.js:492:22)
    at async file:///Users/deno/settimeout_range.test.ts:34:2

 FAILURES 

setTimeout(): Schedules intervals… ... greater than or equal to 2 ** 31 => ./settimeout_range.test.ts:34:10

FAILED | 0 passed (1 step) | 1 failed (1 step) (5ms)

error: Test failed
```

**Comment by iuioiua** [comment]
@jsejcksn, would you be able to submit a PR containing your suggestions?

**Comment by jsejcksn** [comment]
> @jsejcksn, would you be able to submit a PR containing your suggestions?

@iuioiua Created: https://github.com/denoland/std/pull/5834

**Pull request merged**